libdnf/base/transaction.cpp

```diff
@@ -323,13 +273,28 @@ TransactionPackage Transaction::Impl::make_transaction_package(
     Id id,
     TransactionPackage::Action action,
     rpm::solv::GoalPrivate & solved_goal,
-    std::map<Id, std::vector<Id>> & replaced) {
+    std::map<Id, std::vector<Id>> & replaced,
+    rpm::PackageQuery installed_query) {
     auto obs = solved_goal.list_obsoleted_by_package(id);
 
-    libdnf_assert(!obs.empty(), "No obsoletes for {}", transaction_item_action_to_string(action));
-
     rpm::Package new_package(base, rpm::PackageId(id));
-    auto reason = new_package.get_reason();
+
+    transaction::TransactionItemReason reason;
+    if (action == TransactionPackage::Action::INSTALL) {
+        reason = std::max(new_package.get_reason(), solved_goal.get_reason(id));
```

Comment:
This is good and I like it. I believe that `new_package.get_reason()` should return a different reason for not installed packages and for installed packaged installed outside of our stack. Also that would be the best place to put the logic because it is outside of system state but still internal. What do you think?

Reply:
I see what you mean, but I don't like it. The `TransactionItemReason` is already a hodge-podge of values that are only applicable in specific cases and this adds one more that is again special. In addition on a `Package` class level the case (installed outside dnf) is represented by `is_installed()` returning `true` and `get_reason()` returning `None`, so it's not that we don't have this easily covered on the API.

Again, if you know of cases for which the current implementation will fail, let me know so I can add them to the CI stack tests.

Reply:
The first issue with the code is already mentioned bellow - evaluation redundancy, but this is not a problem. I believe that the package `get_reason()` should distinguish between two different reasons that are now represented by the None reason. Not installed or installed outside of libdnf5 stack and also removed outside of libdnf5 stack (present entry in system state but package is not installed anymore). It has also one advantage we do not need to test None reason in code, like it is done on following lines. The lines bellow did not make a decision how None will be handled therefore somewhere else the decision will be made and there will be a suggestion that None means modified outside of DNF. The huge advantage will be for repertory command where installed outside of DNF5 will be explicitly visible as a specific reason.

Reply:
I also think that the code will have a problem with handling of obsoletes on line 303 `if (old_reson > reason) {`.

I took a scenario https://github.com/rpm-software-management/ci-dnf-stack/blob/master/dnf-behave-tests/dnf/obsoletes.feature and I modified install line to install it with rpm.
```
@dnf5
# TODO(nsella) transaction table output
Scenario: Autoremoval of obsoleted package
   When I execute rpm with args "install PackageB-1.0"
   Then the exit code is 0
   When I execute dnf with args "upgrade"
   Then the exit code is 0
    And Transaction is following
        | Action        | Package                                   |
        | install       | PackageB-Obsoleter-0:1.0-1.x86_64         |
        | obsoleted     | PackageB-0:1.0-1.x86_64                   |
   When I execute dnf with args "autoremove"
   Then the exit code is 0
    But Transaction is empty
```
I guess that autoremove will remove obsoleted package as unused dep. It is because libsolv for such a tranaction will return `dep` and `PackageB` installed reason is None and None is lover then Dep. I gues that similar problem is also with upgrade depependency that obsoletes package with None reason.

Reply:
We can also discuss what is expected the return value for available packages for `new_package.get_reason()` - whether it is None or a reason of installed packages with the same NA.

Reply:
There is one additional eddge case with installroot packages. Imagine situation when I will install one installonly package (version 1) (requested by user therefore userinstalled) and the second one will be injected into transaction as a dependency (version 2) or obsoleter or obsoleter package installed outside of DNF5. Then during processing of packages in installset than the order of them will result effect the final reason. It can be resolved by additional sorting of transaction items or during updating System state, but both code requires reason comparison.

Reply:
> I also think that the code will have a problem with handling of obsoletes on line 303 `if (old_reson > reason) {`.
> 
> I took a scenario https://github.com/rpm-software-management/ci-dnf-stack/blob/master/dnf-behave-tests/dnf/obsoletes.feature and I modified install line to install it with rpm.
> 
> ```
> @dnf5
> # TODO(nsella) transaction table output
> Scenario: Autoremoval of obsoleted package
>    When I execute rpm with args "install PackageB-1.0"
>    Then the exit code is 0
>    When I execute dnf with args "upgrade"
>    Then the exit code is 0
>     And Transaction is following
>         | Action        | Package                                   |
>         | install       | PackageB-Obsoleter-0:1.0-1.x86_64         |
>         | obsoleted     | PackageB-0:1.0-1.x86_64                   |
>    When I execute dnf with args "autoremove"
>    Then the exit code is 0
>     But Transaction is empty
> ```
> 
> I guess that autoremove will remove obsoleted package as unused dep. It is because libsolv for such a tranaction will return `dep` and `PackageB` installed reason is None and None is lover then Dep. I gues that similar problem is also with upgrade depependency that obsoletes package with None reason.

Okay, this case is essentially the same as the one I came up with. The reason for this one actually ends up being "User", i.e. the following scenario passes (note the `package state is` step at the end):
```gherkin
@dnf5
Scenario: Autoremoval of obsoleted package
   When I execute rpm with args "-i --nodeps {context.scenario.repos_location}/dnf-ci-obsoletes/x86_64/PackageB-1.0-1.x86_64.rpm"
   Then the exit code is 0
   When I execute dnf with args "upgrade"
   Then the exit code is 0
    And Transaction is following
        | Action        | Package                                   |
        | install       | PackageB-Obsoleter-0:1.0-1.x86_64         |
        | obsoleted     | PackageB-0:1.0-1.x86_64                   |
   When I execute dnf with args "autoremove"
   Then the exit code is 0
    But Transaction is empty
    And package state is
        | package                         | userinstalled | from_repo        |
        | PackageB-Obsoleter-1.0-1.x86_64 | True          | dnf-ci-obsoletes |
```

This is because the action for the package is `INSTALL` and the reason is taken from the goal, which gives `USER` as reason.

I'm thinking of a potential solution to unconditionally force `NONE` as the final reason in case one of the obsoleted packages has reason `NONE`...